### ISSUE TITLE
not showing any notifications

### ISSUE THREAD

**Issue by 7GRAFIX** [ThemeAvenue/Remote-Dashboard-Notifications#7]
Hi,

I have installed the plugin properly, but I'm not able to show any notifications on client site. 
I have added the few lines of code to my child theme functions.php but nothing shows up
I'm also using the correct channel id's and keys. 
But there are also no errors, so I'm quite sure everything fine at client side but not about server side. 

I have created 2 notifications, a few channels on the server site, no post type limitation so they should be visible. 

In your documentation here on github, it says 
$notification = new TAV_Remote_Notification_Client(274,'0f0201703ea31bd6', 'http://url.com**?post_type=notification**');

On your own website is says same without ?post_type=notification in url
$notification = new TAV_Remote_Notification_Client(274,'0f0201703ea31bd6', 'http://url.com');

So which one is correct or do I need to add something more to get this working? 

Thanks!

**Comment by julien731** [comment]
Good point. The docs should be updated. Using the domain name (without the `?` part) is fine. If you've added the `?` part though the script is backwards compatible.

Regarding the issue, can you confirm that you've added the client file `class-remote-notification-client.php` to your theme?

**Comment by 7GRAFIX** [comment]
OK, thanks for the feedback!

Yes, I have added that file into my child theme folder directly and main theme folder but not working. 
Unless it has to be somewhere else specific? /themes/includes/ ??

<img width="529" alt="schermafdruk 2016-04-20 14 41 38" src="https://cloud.githubusercontent.com/assets/7568546/14674872/615a5658-0706-11e6-819f-729a4cfe773e.png">

**Comment by julien731** [comment]
As long as you load the file using `include()` or `require()` there is no specific location. You did load it right?

**Comment by 7GRAFIX** [comment]
Yes, I have following lines in my functions.php in child theme:

<img width="1040" alt="schermafdruk 2016-04-20 14 57 09" src="https://cloud.githubusercontent.com/assets/7568546/14675251/5fa5aa54-0708-11e6-9768-94e7c4034dc9.png">

**Comment by julien731** [comment]
Ok. Sorry to ask the basic questions but I need to make sure the problem is somewhere else ;)

Could you please install [Query Monitor](https://wordpress.org/plugins/query-monitor/) on your site and inspect the HTTP queries (it will display in your admin bar).

- Can you find the query to http://www.7grafix.be?post_type=notification?
- Is the query successful or not (it will show in red if not)?

**Comment by 7GRAFIX** [comment]
No problems with that.
At least, this is the only thing I can find regarding remote-notifications

<img width="1713" alt="schermafdruk 2016-04-20 15 07 36" src="https://cloud.githubusercontent.com/assets/7568546/14675522/acd4ec1c-0709-11e6-80f8-395dc0361a06.png">

**Comment by julien731** [comment]
This means the class is indeed in use. Could you please check the HTTP requests, though. See the image attached.

![image](https://cloud.githubusercontent.com/assets/1441024/14675715/8875e7e2-0734-11e6-8a71-c540eb8088a2.png)

**Comment by 7GRAFIX** [comment]
Strange, that's completely empty at my side while the function is called from withing child theme :/

<img width="1740" alt="schermafdruk 2016-04-20 15 25 28" src="https://cloud.githubusercontent.com/assets/7568546/14676061/2eb8bff4-070c-11e6-9498-f31a36aadfd6.png">

**Comment by julien731** [comment]
That's ok. It just means that there are no HTTP requests done at this particular page load.

Next step, and sorry for yet another plugin, could you:

- Install [WP Sweep](https://wordpress.org/plugins/wp-sweep/)
- Use it to clear the transients (you'll find WP Sweep under *Tools*)
- Once the action is finished, go to any page (the dashboard for instance) and check the HTTP requests again

**Comment by 7GRAFIX** [comment]
Done, http requests are still empty. No records at all.

**Comment by 7GRAFIX** [comment]
This is what's running on server side:
<img width="1448" alt="schermafdruk 2016-04-20 15 49 54" src="https://cloud.githubusercontent.com/assets/7568546/14676887/a88911e6-070f-11e6-88a4-f28a4cdea02f.png">

**Comment by 7GRAFIX** [comment]
Hi Julien, Any idea about this? 
If you add that line with my details, do you get a http req response back or can you see the message in your dashboard? 

Again, I'm using a LEMP stack server with varnish and memcached. Could be it that something from caching is blocking this from server side setup? I do have cleared cache to make sure I have latest info, but still nothing.
I'm also using a security plugin (iThemes Security Pro) and I have blocked xml-rpc completely. 
Not sure if that is interfering, but any other API connection (like Woo iOS app) if working perfectly fine with xml-rpc turned off completely.

Thanks for your feedback!

**Comment by 7GRAFIX** [comment]
Hi Julien,

Any update so far?

**Comment by julien731** [comment]
Hi Fabio. I just released a new version. Would you mind updating and trying with v1.3.0? https://github.com/ThemeAvenue/Remote-Dashboard-Notifications/releases/tag/1.3.0

**Comment by codeagencybe** [comment]
Hi Julien, 

Yes, now the messages are showing but another issue arises.
I have created 2 test messages to check if everything was working. Now I have deleted these 2 messages from server and created 1 new "official" message but all my client sites keep showing the 2 old messages and my new message is not showing. 
I already have clear cache (browser, server - varnish) but the 2 old messages don't go away. 
Also, clicking the "X" button at client side to close the message is not working.

**Comment by julien731** [comment]
That is most likely due to the transient being used for caching things (and reducing the amount of HTTP requests).

In order to clear that, you can use [WP Sweep](https://wordpress.org/plugins/wp-sweep/) on your client sites and use the "Clear Transients" (or something like that) option.

Regarding the notices, not closing, can you confirm that you updated the client class `class-remote-notification-client.php` on your client sites?

**Comment by codeagencybe** [comment]
Hi Julien,

I did clear all the transients (I'm using a different premium plugin but it does the same result). 
Even when I add a NEW client to pull messages from server, it still keeps loading the old messages. 
There can not be any transients for that site, and yet it pulls old data. 

Yes, I have updated also the client class at client website. I have overwritten the file in child theme and clear all cache client side and server side.

**Comment by julien731** [comment]
I'll have to check if there is a problem with the posts order on the server side then. That could explain the fact that you're getting old posts.

For the notice dismissal it is quite weird though. I did test it earlier and it was working. FYI, the dismissal is per user if it makes any difference.

**Comment by codeagencybe** [comment]
I'm going to reboot my VPS server, just to make sure there is no server side caching issue. 
Another thing I noticed, the message does not respect my markup at server side. 

I made a few paragraphs, but it just adds everything after eachother.
Screenshots:

<img width="1734" alt="schermafdruk 2016-04-27 15 58 42" src="https://cloud.githubusercontent.com/assets/7568546/14854686/f178230c-0c90-11e6-93cd-d3323e87aa74.png">
<img width="1425" alt="schermafdruk 2016-04-27 15 58 35" src="https://cloud.githubusercontent.com/assets/7568546/14854687/f17d6cc2-0c90-11e6-8d86-d3027f9568ad.png">

**Comment by codeagencybe** [comment]
Hi Julien,
Now the messages are showing up on some client sites. Not all yet, but perhaps it could be due to cache sync at background. 
I'm using Varnish + memcached at all my vps servers, so it could be the cause for longer displaying the old information. 
Can you tell me if there any specific files that I should never cache? I can make some exclusion rules so new info can popup immediately.

Thanks!

**Comment by julien731** [comment]
Well, the new version makes everything go through an Ajax process so there shouldn't be caching issues. I'm not sure there is something specific to setup here.

Regarding the layout it doesn't look great indeed. I'll have a look at it. Might need to pass the content through [`wpautop()`](https://codex.wordpress.org/Function_Reference/wpautop).

**Comment by codeagencybe** [comment]
Hi Julien,

I have setup a new and seperate "notification server" on a test environment and created a very small client-side plugin to make it more easy to deploy to several client websites via iThemes Sync or mainWP.

The only thing that is still not OK, is the output. It doesn't respect the formatting from serverside editor. 
Also, the title is not visible at client side. Somewhere it's forgotten. It starts immediately with the content. 

Side note: can your plugin also handle rich content/media?
E.g., let's assume I put some Youtube video code at server server to show a simple tutorial to explain a new feature, will it also be available at client side? 
What about a contact7 form? If I put in the shortcode at server side to ask e.g. a survey about customer satisfaction, will the form be working nicely at client side? 
This could bring a lot more possibilities like showing a download URL to some PDF file (from server side), showing an image(s), ...

**Comment by codeagencybe** [comment]
Hello Julien,

Is it possible to get the channel ID and key via the client automatically based on the domain name? 
I'll explain why:

I have 3 channels:
2 channels are "global" for all client sites (to show general scheduled maintenance and server maintenance) and 1 is client specific to show messages only for that client.

At server side I'm using the domainname as channel name like:
client1domainname.com
client2domainname.com
etc...

It would be interesting to send a a req with domain name from client site and do something like this in the client side plugin:

//Get rid of www
$client_domain_name =  preg_replace('/^www./','',$_SERVER['SERVER_NAME']);

require( 'class-remote-notification-client.php' );
if ( function_exists( 'rdnc_add_notification' ) ) {
    //code here to collect channel ID and key for $client_domain_name
    $client_channel_ID = ????
    $client_channel_key = ?????
    rdnc_add_notification( $client_channel_ID, $client_channel_key, 'http://myupdateserver.com' ); 
}

What do you think?

**Comment by julien731** [comment]
Regarding formatting, I created #10 

Regarding rich content, this is not something we will implement. This plugin was only designed to give short notifications to a client site without altering the client's experience. We believe that adding rich content would start to degrade the experience.

Finally, regarding the channel ID & key, this will not be possible either. The channel ID is the one constant that allows fetching the channel, and the key is a (simple) authentication key, which means that it shouldn't be given by the server or it loses all its purpose.

Even if we were to automatically fetch the channel ID only, it wouldn't be safe enough because a channel name can be changed. We absolutely need a constant here, and only the channel ID is a constant in this scenario.

**Comment by codeagencybe** [comment]
Hi Julien,

Thanks! Looking forward for the solution so the formatting is right. 

Yes, regarding rich content you are right. It's off-track from the original idea/purpose for this plugin.
But I like the idea though for such plugin than can distribute content directly into client website dashboard. It could also serve as some kind of education/tutorial system for new clients rolling into WP. 
Perhaps this can be a foundation for a new idea/plugin in the future. 
For now, I focus on the simple notifications to alert clients about upcoming maintenance and updates.

About channel ID & key, you are right too. I didn't think of the security hole for that. 
I'm going to try to create an admin page for my client plugin to manage the channels via WP backend instead of modifying the plugin file. 
2 channels will be default always included hard coded, the other one (or even extra's) can be added via backend management page by filling in a 2 field form for channel ID and key and stored in DB.
